File: src/services/cloudwatch.js
// services/cloudwatch.js — CloudWatch metrics. The AWS SDK v3 CloudWatch client
// speaks awsJson1.0 with X-Amz-Target: GraniteServiceVersion20100801.<Op>.
// Backs onto the in-memory ring buffer in store.js (store.cloudwatch.metrics +
// putMetric). Distinct from routes/cloudwatch.js, which is the UI plane.
import { store } from '../store.js';
import { jsonResponse, errorJson } from '../middleware/response.js';

export function handler(req, res) {
  const op   = (req.headers['x-amz-target'] || '').split('.')[1] || '';
  const body = req.parsedBody || {};
  switch (op) {
    case 'PutMetricData':       return putMetricData(res, body);
    case 'GetMetricStatistics': return getMetricStatistics(res, body);
    case 'ListMetrics':         return listMetrics(res, body);
    default:                    return errorJson(res, 400, 'UnknownOperationException', `Unknown CloudWatch op: ${op}`);
  }
}

function putMetricData(res, body) {
  const ns = body.Namespace || '';
  for (const d of body.MetricData || []) {
    store.putMetric(ns, d.MetricName, Number(d.Value ?? 0), d.Unit || 'None');
  }
  return jsonResponse(res, 200, {});
}

function getMetricStatistics(res, body) {
  const points = store.cloudwatch.metrics[`${body.Namespace}/${body.MetricName}`] || [];
  const start  = toMs(body.StartTime) || 0;
  const end    = toMs(body.EndTime) || Date.now();
  const period = (Number(body.Period) || 60) * 1000;
  const stats  = body.Statistics && body.Statistics.length ? body.Statistics : ['Average'];

  const buckets = new Map();
  let unit = 'None';
  for (const p of points) {
    if (p.t < start || p.t > end) continue;
    unit = p.unit || unit;
    const b = Math.floor(p.t / period) * period;
    (buckets.get(b) || buckets.set(b, []).get(b)).push(p.v);
  }
  const Datapoints = [...buckets.entries()].sort((a, b) => a[0] - b[0]).map(([b, vals]) => {
    const dp = { Timestamp: Math.floor(b / 1000), Unit: unit };   // awsJson timestamps are epoch seconds
    if (stats.includes('Sum'))         dp.Sum = vals.reduce((a, c) => a + c, 0);
    if (stats.includes('Average'))     dp.Average = vals.reduce((a, c) => a + c, 0) / vals.length;
    if (stats.includes('Minimum'))     dp.Minimum = Math.min(...vals);
    if (stats.includes('Maximum'))     dp.Maximum = Math.max(...vals);
    if (stats.includes('SampleCount')) dp.SampleCount = vals.length;
    return dp;
  });
  return jsonResponse(res, 200, { Label: body.MetricName, Datapoints });
}

function listMetrics(res, body) {
  const Metrics = Object.keys(store.cloudwatch.metrics).map(key => {
    const i = key.indexOf('/');
    return { Namespace: key.slice(0, i), MetricName: key.slice(i + 1), Dimensions: [] };
  }).filter(m => !body.Namespace || m.Namespace === body.Namespace);
  return jsonResponse(res, 200, { Metrics });
}

// awsJson1.0 serializes timestamps as epoch seconds (number).
function toMs(t) {
  if (t == null) return 0;
  if (typeof t === 'number') return t * 1000;
  const n = Number(t);
  return Number.isNaN(n) ? (Date.parse(t) || 0) : n * 1000;
}

File: internal/services/cloudwatch/cloudwatch.go
// Package cloudwatch — port of src/services/cloudwatch.js (awsJson1.0,
// X-Amz-Target: GraniteServiceVersion20100801.<Op>). Backs onto the metrics
// ring buffer in state (PutMetric).
package cloudwatch

import (
	"encoding/json"
	"math"
	"net/http"
	"sort"
	"strings"
	"time"

	"github.com/mockcloud/mockcloud/internal/httpapi"
	"github.com/mockcloud/mockcloud/internal/protocol/respond"
	"github.com/mockcloud/mockcloud/internal/state"
	"github.com/mockcloud/mockcloud/internal/store"
)

func Handler(w http.ResponseWriter, r *httpapi.Request, st *store.Store) {
	target := r.Header.Get("x-amz-target")
	op := ""
	if i := strings.Index(target, "."); i >= 0 {
		op = target[i+1:]
	}
	b := r.ParsedBody
	switch op {
	case "PutMetricData":
		putMetricData(w, b, st)
	case "GetMetricStatistics":
		getMetricStatistics(w, b, st)
	case "ListMetrics":
		listMetrics(w, b, st)
	default:
		respond.ErrorJSON(w, 400, "UnknownOperationException", "Unknown CloudWatch op: "+op)
	}
}

func putMetricData(w http.ResponseWriter, b map[string]any, st *store.Store) {
	ns := httpapi.Str(b, "Namespace")
	st.With(func(s *state.State) {
		data, _ := b["MetricData"].([]any)
		for _, d := range data {
			dm, ok := d.(map[string]any)
			if !ok {
				continue
			}
			value, _ := httpapi.Num(dm, "Value") // Number(d.Value ?? 0)
			unit := httpapi.Str(dm, "Unit")
			if unit == "" {
				unit = "None"
			}
			s.PutMetric(ns, httpapi.Str(dm, "MetricName"), value, unit)
		}
	})
	respond.JSON(w, 200, map[string]any{})
}

// toMs — awsJson1.0 serializes timestamps as epoch seconds (number); string
// forms fall back to numeric-then-date parsing (src/services/cloudwatch.js).
func toMs(v any) int64 {
	switch t := v.(type) {
	case nil:
		return 0
	case json.Number:
		f, err := t.Float64()
		if err != nil {
			return 0
		}
		return int64(f * 1000)
	case float64:
		return int64(t * 1000)
	case string:
		if parsed, err := time.Parse(time.RFC3339, t); err == nil {
			return parsed.UnixMilli()
		}
		return 0
	}
	return 0
}

func getMetricStatistics(w http.ResponseWriter, b map[string]any, st *store.Store) {
	key := httpapi.Str(b, "Namespace") + "/" + httpapi.Str(b, "MetricName")
	start := toMs(b["StartTime"])
	end := toMs(b["EndTime"])
	if end == 0 {
		end = state.NowMs()
	}
	periodSec, ok := httpapi.Num(b, "Period")
	if !ok || periodSec == 0 {
		periodSec = 60
	}
	period := int64(periodSec) * 1000
	var stats []string
	if raw, ok := b["Statistics"].([]any); ok {
		for _, s := range raw {
			if sv, ok := s.(string); ok {
				stats = append(stats, sv)
			}
		}
	}
	if len(stats) == 0 {
		stats = []string{"Average"}
	}
	has := func(name string) bool {
		for _, s := range stats {
			if s == name {
				return true
			}
		}
		return false
	}

	type bucket struct {
		t    int64
		vals []float64
	}
	var points []state.MetricPoint
	st.With(func(s *state.State) {
		points = append(points, s.CloudWatch.Metrics[key]...)
	})
	buckets := map[int64]*bucket{}
	unit := "None"
	for _, p := range points {
		if p.T < start || p.T > end {
			continue
		}
		if p.Unit != "" {
			unit = p.Unit
		}
		bt := (p.T / period) * period
		if buckets[bt] == nil {
			buckets[bt] = &bucket{t: bt}
		}
		buckets[bt].vals = append(buckets[bt].vals, p.V)
	}
	ordered := make([]*bucket, 0, len(buckets))
	for _, bk := range buckets {
		ordered = append(ordered, bk)
	}
	sort.Slice(ordered, func(i, j int) bool { return ordered[i].t < ordered[j].t })

	datapoints := make([]map[string]any, 0, len(ordered))
	for _, bk := range ordered {
		dp := map[string]any{"Timestamp": bk.t / 1000, "Unit": unit} // epoch seconds
		sum := 0.0
		min, max := math.Inf(1), math.Inf(-1)
		for _, v := range bk.vals {
			sum += v
			if v < min {
				min = v
			}
			if v > max {
				max = v
			}
		}
		if has("Sum") {
			dp["Sum"] = sum
		}
		if has("Average") {
			dp["Average"] = sum / float64(len(bk.vals))
		}
		if has("Minimum") {
			dp["Minimum"] = min
		}
		if has("Maximum") {
			dp["Maximum"] = max
		}
		if has("SampleCount") {
			dp["SampleCount"] = len(bk.vals)
		}
		datapoints = append(datapoints, dp)
	}
	respond.JSON(w, 200, map[string]any{"Label": httpapi.Str(b, "MetricName"), "Datapoints": datapoints})
}

func listMetrics(w http.ResponseWriter, b map[string]any, st *store.Store) {
	wantNs := httpapi.Str(b, "Namespace")
	var metrics []map[string]any
	st.With(func(s *state.State) {
		for key := range s.CloudWatch.Metrics {
			i := strings.Index(key, "/") // FIRST slash — namespaces containing '/' split here (Node quirk)
			ns, name := key[:i], key[i+1:]
			if wantNs != "" && ns != wantNs {
				continue
			}
			metrics = append(metrics, map[string]any{
				"Namespace": ns, "MetricName": name, "Dimensions": []any{},
			})
		}
	})
	sort.Slice(metrics, func(i, j int) bool {
		return metrics[i]["MetricName"].(string) < metrics[j]["MetricName"].(string)
	})
	if metrics == nil {
		metrics = []map[string]any{}
	}
	respond.JSON(w, 200, map[string]any{"Metrics": metrics})
}
